### ISSUE TITLE
Should some kernels not be lazily evaluated?

### ISSUE THREAD

**Issue by jacobrgardner** [cornellius-gp/gpytorch#219]
#217 patches in a kind of gross way an issue where, because we lazily evaluate kernels, we are always calling the default `LazyVariable.exact_predictive*`, because `full_covar` is always a `LazyEvaluatedKernelVariable` by the time we reach here: https://github.com/cornellius-gp/gpytorch/blob/1c8f891981a3d50e57017c7d53f25f7992f000a3/gpytorch/functions/__init__.py#L88

A potential alternative fix would be to have `GridInterpolationKernel` override `__call__` to always directly go to forward and not lazily evaluate anything.

The advantage of this is cleanliness, and perhaps slightly improved performance since we now won't make multiple calls to `GridInterpolationKernel` to get all the matrices involved. 

The disadvantage of this that I can see is that if ever someone wanted to write a LazyVariable that benefits both from custom prediction code AND lazily evaluation the test-train matrices, there will still be no way to do this without a fix like #217.

Thoughts?

**Comment by gpleiss** [comment]
A couple thoughts about this:

- I've never really liked how `exact_predictive_*` lives on the LazyVariable. From a strict OO programming perspective, a LazyVariable shouldn't know about the kind of inference being performed on it. Put a different way, InterpolatedLazyVariables have knowledge of KISS-GP inference (due to the `exact_predictive_*`) methods, yet InterpolatedLazyVariables are strictly more general than KISS-GP, and so this seems like a weird tight coupling.

- That being said, I think back in the day we tried adding the `exact_predictive_*` methods on a separate object, and that ended up being gross as well. And I don't think it ended up being a better separation of concerns in the long run.

- InterpolatedLazyVariable is such a strange beast. It basically overrides every single default method, and it's really one of the only instances where we have special inference rules.

- InterpolatedLazyVariable kernels probably don't need to be lazily evaluated. Lazily evaluating any portion of these kernels doesn't give you any cost savings, since you'll have to do interpolation with the base lazy variable anyways. So in this particular case, non-lazy kernel evaluation makes sense.

**TLDR**

I think the hack you have in #217, or overriding `GridInterpolationKernel.__call__` are both reasonable solutions. In the long term we should think about how to better separate these inference decisions. But for now the hacks are probably reasonable.

**Comment by gpleiss** [comment]
Based on what we've talked about w.r.t. this, I think all kernels should probably be lazily evaluated. We might want to do some refactoring/cleanup tho
